<?xml version="1.0" encoding="UTF-8"?>
<ui version="4.0">
 <class>ConfigureGraphics</class>
 <widget class="QWidget" name="ConfigureGraphics">
  <property name="geometry">
   <rect>
    <x>0</x>
    <y>0</y>
    <width>400</width>
    <height>509</height>
   </rect>
  </property>
  <property name="minimumSize">
   <size>
    <width>0</width>
    <height>0</height>
   </size>
  </property>
  <property name="windowTitle">
   <string>Form</string>
  </property>
  <property name="accessibleName">
   <string>Graphics</string>
  </property>
  <layout class="QVBoxLayout" name="verticalLayout_1">
   <item>
    <widget class="QGroupBox" name="apiBox">
     <property name="title">
      <string>API Settings</string>
     </property>
     <layout class="QVBoxLayout" name="verticalLayout_2">
      <item>
       <widget class="QWidget" name="graphics_api_group" native="true">
        <layout class="QHBoxLayout" name="graphics_api_group_2">
         <property name="leftMargin">
          <number>0</number>
         </property>
         <property name="topMargin">
          <number>0</number>
         </property>
         <property name="rightMargin">
          <number>0</number>
         </property>
         <property name="bottomMargin">
          <number>0</number>
         </property>
         <item>
          <widget class="QLabel" name="graphics_api_label">
           <property name="text">
            <string>Graphics API</string>
           </property>
          </widget>
         </item>
         <item>
          <widget class="QComboBox" name="graphics_api_combo">
           <item>
            <property name="text">
             <string>Software</string>
            </property>
           </item>
           <item>
            <property name="text">
             <string>OpenGL (default)</string>
            </property>
            <property name="toolTip">
             <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Use OpenGL renderer. Minimum requirement for GPU support is OpenGL 4.3.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
            </property>
           </item>
           <item>
            <property name="text">
             <string>Vulkan</string>
            </property>
            <property name="toolTip">
             <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Use Vulkan renderer. Minimum requirement for GPU support is Vulkan 1.2. Best results with Vulkan 1.3 or greater.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
            </property>
           </item>
          </widget>
         </item>
        </layout>
       </widget>
      </item>
      <item>
       <widget class="QWidget" name="physical_device_group" native="true">
        <layout class="QHBoxLayout" name="physical_device_group_2">
         <property name="leftMargin">
          <number>0</number>
         </property>
         <property name="topMargin">
          <number>0</number>
         </property>
         <property name="rightMargin">
          <number>0</number>
         </property>
         <property name="bottomMargin">
          <number>0</number>
         </property>
         <item>
          <widget class="QLabel" name="physical_device_label">
           <property name="text">
            <string>Physical Device</string>
           </property>
          </widget>
         </item>
         <item>
          <widget class="QComboBox" name="physical_device_combo"/>
         </item>
        </layout>
       </widget>
      </item>
      <item>
       <widget class="QWidget" name="opengl_renderer_group" native="true">
        <layout class="QGridLayout" name="opengl_renderer_group_2">
         <property name="leftMargin">
          <number>0</number>
         </property>
         <property name="topMargin">
          <number>0</number>
         </property>
         <property name="rightMargin">
          <number>0</number>
         </property>
         <property name="bottomMargin">
          <number>0</number>
         </property>
         <item row="0" column="0">
          <widget class="QLabel" name="opengl_renderer_label">
           <property name="text">
            <string>OpenGL Renderer</string>
           </property>
          </widget>
         </item>
         <item row="0" column="1">
          <widget class="QLabel" name="opengl_renderer_name_label"/>
         </item>
         <item row="1" column="0">
          <widget class="QCheckBox" name="toggle_use_gles">
           <property name="toolTip">
            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Use OpenGLES extensions instead of regular OpenGL. Primarily for use with GPUs on embedded systems. Minimum requirement for GPU support is OpenGLES 3.1. Best results with OpenGLES 3.2 or greater.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
           </property>
           <property name="text">
            <string>Use GLES extensions</string>
           </property>
          </widget>
         </item>
        </layout>
       </widget>
      </item>
      <item>
       <layout class="QGridLayout" name="gridLayout_1">
        <item row="0" column="0">
         <widget class="QCheckBox" name="spirv_shader_gen">
          <property name="toolTip">
           <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Emits the fragment shader used to emulate PICA using SPIR-V instead of GLSL.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
          </property>
          <property name="text">
           <string>SPIR-V Shader Generation</string>
          </property>
         </widget>
        </item>
        <item row="0" column="1">
         <widget class="QCheckBox" name="toggle_geometry_shader">
          <property name="toolTip">
           <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Enable Geometry Shaders. Improved accuracy but extremely expensive on tilers.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
          </property>
          <property name="text">
           <string>Geometry Shader</string>
          </property>
         </widget>
        </item>
        <item row="0" column="2">
         <widget class="QCheckBox" name="toggle_relaxed_precision">
          <property name="toolTip">
           <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Optimizes shader code by allowing certain floating-point operations to use reduced precision. Can improve performance by reducing memory bandwidth and computational requirements. May cause issues on certain GPUs.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
          </property>
          <property name="text">
           <string>Relaxed Precision</string>
          </property>
         </widget>
        </item>
        <item row="0" column="3">
         <widget class="QCheckBox" name="sample_shading">
          <property name="toolTip">
           <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Enables a Vulkan extension that may improve the rendering quality, at the cost of some performance.&lt;/p&gt;&lt;p&gt;Requires a relatively powerful GPU for better performance.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
          </property>
          <property name="text">
           <string>Sample Shading</string>
          </property>
         </widget>
        </item>
       </layout>
      </item>
      <item>
       <widget class="QGroupBox" name="widget_optimize_spirv">
        <property name="minimumSize">
         <size>
          <width>0</width>
          <height>0</height>
         </size>
        </property>
        <property name="title">
         <string>SPIR-V Shader Optimization</string>
        </property>
        <layout class="QHBoxLayout" name="verticalLayout_3">
         <item> 
          <layout class="QHBoxLayout" name="horizontalLayout_1">
           <item>
            <widget class="QLabel" name="label_optimize_spirv">
             <property name="toolTip">
              <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Performs various optimizations on SPIR-V modules to improve performance and reduce size. Equivalent to passing -O or -Os to spirv-opt.&lt;/p&gt;&lt;p&gt;NOTE: To prevent visual artifacts, make sure to empty the shader cache in the Borked3DS user folder before changing optimization level. The Borked3DS user folder can be accessed through the File menu.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
             </property>
             <property name="text">
               <string>Optimization Level</string>
             </property>
            </widget>
           </item>
           <item>
            <widget class="QComboBox" name="optimize_spirv_combobox">
             <item>
              <property name="text">
                  <string>Disabled</string>
              </property>
             </item>
             <item>
              <property name="text">
                  <string>Performance</string>
              </property>
             </item>
             <item>
              <property name="text">
                  <string>Size</string>
              </property>
             </item>
            </widget>
           </item>
          </layout>
         </item>
         <item>
          <layout class="QHBoxLayout" name="horizontalLayout_2">
           <item>
            <widget class="QCheckBox" name="toggle_spirv_validation">
             <property name="toolTip">
              <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Activate SPIR-V output validation. The validator checks validation rules described by the SPIR-V specification.&lt;/p&gt;&lt;p&gt;Ensures that outputs are valid and protects from bad inputs. Useful for debugging.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
             </property>
             <property name="text">
              <string>Shader Validation</string>
             </property>
            </widget>
           </item>
           <item>
            <widget class="QCheckBox" name="toggle_spirv_legalization">
             <property name="toolTip">
              <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Activate SPIR-V output legalization. Legalizes the SPIR-V code for Vulkan. Primarily used for HLSL shaders.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
             </property>
             <property name="text">
              <string>Shader Legalization</string>
             </property>
            </widget>
           </item>
          </layout>
         </item>
        </layout>
       </widget>
      </item>
     </layout>
    </widget>
   </item>
   <item>
    <widget class="QGroupBox" name="rendererBox">
     <property name="title">
      <string>Renderer</string>
     </property>
     <layout class="QVBoxLayout" name="verticalLayout_4">
      <item>
       <widget class="QWidget" name="hw_renderer_group" native="true">
        <layout class="QGridLayout" name="gridLayout_2">
         <property name="leftMargin">
          <number>0</number>
         </property>
         <property name="topMargin">
          <number>0</number>
         </property>
         <property name="rightMargin">
          <number>0</number>
         </property>
         <property name="bottomMargin">
          <number>0</number>
         </property>
         <item row="1" column="0">
          <widget class="QCheckBox" name="toggle_hw_shader">
           <property name="toolTip">
            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Use the selected graphics API to accelerate shader emulation.&lt;/p&gt;&lt;p&gt;Requires a relatively powerful GPU for better performance.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
           </property>
           <property name="text">
            <string>Hardware Shader Acceleration</string>
           </property>
          </widget>
         </item>
         <item row="2" column="0">
          <widget class="QWidget" name="hw_shader_group" native="true">
           <layout class="QVBoxLayout" name="verticalLayout_5">
            <property name="leftMargin">
             <number>16</number>
            </property>
            <property name="topMargin">
             <number>0</number>
            </property>
            <property name="rightMargin">
             <number>0</number>
            </property>
            <property name="bottomMargin">
             <number>0</number>
            </property>
            <item>
             <widget class="QCheckBox" name="toggle_accurate_mul">
              <property name="toolTip">
               <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Correctly handle all edge cases in multiplication operation in shaders. &lt;/p&gt;&lt;p&gt;Some games requires this to be enabled for the hardware shader to render properly.&lt;/p&gt;&lt;p&gt;However this would reduce performance in most games.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
              </property>
              <property name="text">
               <string>Accurate Multiplication</string>
              </property>
             </widget>
            </item>
           </layout>
          </widget>
         </item>
         <item row="1" column="1">
          <widget class="QCheckBox" name="toggle_shader_jit">
           <property name="toolTip">
            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Use the JIT engine instead of the interpreter for software shader emulation. &lt;/p&gt;&lt;p&gt;Enable this for better performance.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
           </property>
           <property name="text">
            <string>Shader JIT Engine</string>
           </property>
          </widget>
         </item>
         <item row="2" column="1">
          <widget class="QCheckBox" name="toggle_async_shaders">
           <property name="toolTip">
            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Compiles shaders in the background to reduce stuttering during gameplay. When enabled, expect temporary graphical glitches.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
           </property>
           <property name="text">
            <string>Async Shader Compilation</string>
           </property>
          </widget>
         </item>
         <item row="1" column="3">
          <widget class="QCheckBox" name="toggle_async_present">
           <property name="toolTip">
            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Perform presentation on separate threads. Improves performance when using Vulkan in most games.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
           </property>
           <property name="text">
            <string>Async Presentation</string>
           </property>
          </widget>
         </item>
        </layout>
       </widget>
      </item>
     </layout>
    </widget>
   </item>
   <item>
    <widget class="QGroupBox" name="tweaksBox">
     <property name="title">
      <string>Graphics Hacks (Experimental)</string>
     </property>
     <layout class="QVBoxLayout" name="verticalLayout_6">
      <item>
       <widget class="QWidget" name="tweaksGrid" native="true">
        <layout class="QGridLayout" name="gridLayout_3">
         <property name="leftMargin">
          <number>0</number>
         </property>
         <property name="topMargin">
          <number>0</number>
         </property>
         <property name="rightMargin">
          <number>0</number>
         </property>
         <property name="bottomMargin">
          <number>0</number>
         </property>
         <item row="0" column="0">
          <widget class="QCheckBox" name="toggle_skip_slow_draw">
           <property name="toolTip">
            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Skips the slow drawing event from the PICA core. May improve performance, but can cause visual artifacts.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
           </property>
           <property name="text">
            <string>Skip Slow Draw</string>
           </property>
          </widget>
         </item>
         <item row="0" column="1">
          <widget class="QCheckBox" name="toggle_skip_cpu_write">
           <property name="toolTip">
            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Skips the CPU write event from rasterizer cache invalidation. May improve performance but may cause visual artifacts.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
           </property>
           <property name="text">
            <string>Skip CPU Write</string>
           </property>
          </widget>
         </item>
         <item row="0" column="2">
          <widget class="QCheckBox" name="toggle_skip_texture_copy">
           <property name="toolTip">
            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Skips the texture copy event from the rasterizer cache. May improve performance, but can cause visual artifacts.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
           </property>
           <property name="text">
            <string>Skip Texture Copy</string>
           </property>
          </widget>
         </item>
         <item row="0" column="3">
          <widget class="QCheckBox" name="toggle_upscaling_hack">
           <property name="toolTip">
            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Overrides upscaling for dst_params. May fix texture ghosting for some games, but may crash other games as well.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
           </property>
           <property name="text">
            <string>Upscaling Hack</string>
           </property>
          </widget>
         </item>
        </layout>
       </widget>
      </item>
     </layout>
    </widget>
   </item>
   <item>
    <widget class="QGroupBox" name="groupBox">
     <property name="title">
      <string>Advanced</string>
     </property>
     <layout class="QVBoxLayout" name="verticalLayout_7">
      <item>
       <widget class="QWidget" name="widget_texture_sampling" native="true">
        <layout class="QHBoxLayout" name="horizontalLayout_3">
         <property name="leftMargin">
          <number>0</number>
         </property>
         <property name="topMargin">
          <number>0</number>
         </property>
         <property name="rightMargin">
          <number>0</number>
         </property>
         <property name="bottomMargin">
          <number>0</number>
         </property>
         <item>
          <widget class="QLabel" name="texture_sampling_label">
           <property name="toolTip">
            <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Overrides the sampling filter used by games. This can be useful in certain cases with poorly behaved games when upscaling. If unsure set this to Game Controlled.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
           </property>
           <property name="text">
            <string>Texture Sampling</string>
           </property>
          </widget>
         </item>
         <item>
          <widget class="QComboBox" name="texture_sampling_combobox">
           <item>
            <property name="text">
             <string>Game Controlled (default)</string>
            </property>
           </item>
           <item>
            <property name="text">
             <string>Nearest Neighbor</string>
            </property>
           </item>
           <item>
            <property name="text">
             <string>Linear</string>
            </property>
           </item>
          </widget>
         </item>
        </layout>
       </widget>
      </item>
      <item>
       <layout class="QGridLayout" name="gridLayout_4">
        <item row="1" column="0">
         <widget class="QCheckBox" name="toggle_disk_shader_cache">
          <property name="toolTip">
           <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Reduce stuttering by storing and loading generated shaders to disk. It cannot be used without Hardware Shader Acceleration.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
          </property>
          <property name="text">
           <string>Use Disk Shader Cache</string>
          </property>
         </widget>
        </item>
        <item row="1" column="1">
         <widget class="QCheckBox" name="toggle_vsync_new">
          <property name="toolTip">
           <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;VSync prevents the screen from tearing, but some graphics cards have lower performance with VSync enabled. Keep it enabled if you don't notice a performance difference.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
          </property>
          <property name="text">
           <string>Enable VSync</string>
          </property>
         </widget>
        </item>
       </layout>
      </item>
      <item>
        <widget class="QWidget" name="delay_render_layout" native="true">
          <layout class="QHBoxLayout" name="delay_render_layout_inner">
            <property name="leftMargin">
              <number>0</number>
            </property>
            <property name="topMargin">
              <number>0</number>
            </property>
            <property name="rightMargin">
              <number>0</number>
            </property>
            <property name="bottomMargin">
              <number>0</number>
            </property>
            <item>
              <widget class="QComboBox" name="delay_render_combo">
                <item>
                  <property name="text">
                    <string>Use global</string>
                  </property>
                </item>
                <item>
                  <property name="text">
                    <string>Use per-game</string>
                  </property>
                </item>
              </widget>
            </item>
            <item>
              <widget class="QLabel" name="label_delay_render">
                <property name="text">
                  <string>Delay Game Render Thread</string>
                </property>
                <property name="toolTip">
                  <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Delays the emulated game render thread the specified amount of milliseconds every time it submits render commands to the GPU.&lt;/p&gt;&lt;p&gt;Adjust this feature in the (very few) dynamic-fps games to fix performance issues.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
                </property>
              </widget>
            </item>
            <item>
              <widget class="QSlider" name="delay_render_slider">
                <property name="minimum">
                  <number>0</number>
                </property>
                <property name="maximum">
                  <number>16000</number>
                </property>
                <property name="singleStep">
                  <number>100</number>
                </property>
                <property name="pageStep">
                  <number>250</number>
                </property>
                <property name="value">
                  <number>0</number>
                </property>
                <property name="orientation">
                  <enum>Qt::Horizontal</enum>
                </property>
                <property name="tickPosition">
                  <enum>QSlider::TicksBelow</enum>
                </property>
              </widget>
            </item>
            <item>
              <widget class="QLabel" name="delay_render_display_label">
                <property name="text">
                  <string/>
                </property>
                <property name="alignment">
                  <set>Qt::AlignRight|Qt::AlignTrailing|Qt::AlignVCenter</set>
                </property>
              </widget>
            </item>
          </layout>
        </widget>
      </item>
     </layout>
    </widget>
   </item>
   <item>
    <spacer name="verticalSpacer">
     <property name="orientation">
      <enum>Qt::Vertical</enum>
     </property>
     <property name="sizeHint" stdset="0">
      <size>
       <width>20</width>
       <height>270</height>
      </size>
     </property>
    </spacer>
   </item>
  </layout>
 </widget>
 <tabstops>
  <tabstop>toggle_accurate_mul</tabstop>
  <tabstop>toggle_shader_jit</tabstop>
  <tabstop>toggle_disk_shader_cache</tabstop>
  <tabstop>toggle_vsync_new</tabstop>
 </tabstops>
 <resources/>
 <connections/>
</ui>
